### PR TITLE
Make ContentType(string) constructor throw FormatException instead of IndexOutOfRangeException

### LINKED ISSUE #39337: System.Net.Mime ContentType(string) constructor throws IndexOutOfRangeException
### Description

While initializing a new instance of the System.Net.Mime.ContentType class using a string, I got IndexOutOfRangeException instead of FormatException.

```c#
ContentType ct = new ContentType("application/");
```

According to the documentation, we should only get these 3 exceptions:
https://docs.microsoft.com/en-us/dotnet/api/system.net.mime.contenttype.-ctor?view=netcore-3.1#System_Net_Mime_ContentType__ctor_System_String_ 

### Other information

Stack Trace:
 ```
 Message: 
    System.IndexOutOfRangeException : Index was outside the bounds of the array.
  Stack Trace: 
    MailBnfHelper.ReadToken(String data, Int32& offset, StringBuilder builder)
    ContentType.ParseValue()
    ContentType.ctor(String contentType)
```

### PR DESCRIPTION
**Pull request by avanigupta**
Fix #39337

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=39529) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=39529) before we can accept your contribution.<br/><br/>:x: avanigupta [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=39529)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=39529) it.</sub>

**Comment by avanigupta** [comment]
The build is failing due to this error:
##[error]src/libraries/Common/src/System/Net/Mail/MailBnfHelper.cs(272,46): error CS0117: (NETCORE_ENGINEERING_TELEMETRY=Build) **'SR' does not contain a definition for 'ContentTypeInvalid'**

I'm not sure how to fix this. "ContentTypeInvalid" is defined in the same class "SR" as other error messages being accessed in MailBnfHelper. Visual Studio is not helpful in debugging this because it doesn't throw this error. 
Could you please help in fixing this error?

**Comment by danmosemsft** [comment]
Hi @avanigupta thie file is used in several projects (that's why it's in /common/). The resx file in both those projects needs to include the string. Compare MailHeaderFieldInvalidCharacter which is another string in the same .cs file:

```
C:\git\runtime\src\libraries\System.Net.Http\src\Resources\Strings.resx:
  265      <value>The specified value must be greater than {0}.</value>
  266    </data>
  267:   <data name="MailHeaderFieldInvalidCharacter" xml:space="preserve">
  268      <value>An invalid character was found in the mail header: '{0}'.</value>
  269    </data>

C:\git\runtime\src\libraries\System.Net.Mail\src\Resources\Strings.resx:
  284      <value>This operation cannot be performed on a completed asynchronous result object.</value>
  285    </data>
  286:   <data name="MailHeaderFieldInvalidCharacter" xml:space="preserve">
  287      <value>An invalid character was found in the mail header: '{0}'.</value>
  288    </data>

2 matches across 2 files


Searching 181 files for "ContentTypeInvalid" (regex)

C:\git\runtime\src\libraries\System.Net.Mail\src\Resources\Strings.resx:
  146      <value>An invalid character was found in header name.</value>
  147    </data>
  148:   <data name="ContentTypeInvalid" xml:space="preserve">
  149      <value>The specified content type is invalid.</value>
  150    </data>
```

**Comment by avanigupta** [comment]
Hi @danmosemsft, this PR is ready for review. Thanks!

**Comment by danmosemsft** [comment]
@dotnet/ncl could someone please review?

File: src/libraries/System.Net.Http/src/Resources/Strings.resx
Comment (karelz) [inline src/libraries/System.Net.Http/src/Resources/Strings.resx:267]: Why do we need new entry here? Cannot we repurpose the one thrown for `textplain`?
If that one is not appropriate, then should we migrate it to this one as well?

Also, the error seems to be Mail specific - other entries start with `MailHeader`/`MailAddress` - if there is good reason to keep the new entry, shouldn't we unify the name?

Comment (karelz) [reply src/libraries/System.Net.Http/src/Resources/Strings.resx:267]: Ah, ok - it is just copy of `ContentTypeInvalid` into Common, that makes sense now.

File: src/libraries/Common/src/System/Net/Mail/MailBnfHelper.cs
Comment (karelz) [inline src/libraries/Common/src/System/Net/Mail/MailBnfHelper.cs:272]: What if we return either empty or `null` string in this case? ... Rather than embed semantic knowledge into token parser?

One usage can deal with it:
https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/libraries/System.Net.Mail/src/System/Net/Mime/ContentType.cs#L265-L270

Another usage cannot (I assume it will throw `IndexOutOfRangeException`) - maybe we should fix that rather than proposed change?
https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/libraries/System.Net.Mail/src/System/Net/Mime/ContentDisposition.cs#L302-L309

**Review by karelz — CHANGES_REQUESTED** [review]

File: src/libraries/Common/src/System/Net/Mail/MailBnfHelper.cs
Comment (avanigupta) [reply src/libraries/Common/src/System/Net/Mail/MailBnfHelper.cs:272]: Both ReadParameterAttribute and ReadToken are being invoked in various places. If we stop throwing an exception here, it might not be a trivial change and I'm not sure if the tests are good enough and cover every edge case. I am not too familiar with all the parsing rules so I don't want to unintentionally introduce a breaking change. 

But I agree that token parser doesn't need to know about content type semantics. How about instead of "ContentTypeInvalid" we throw a slightly more generic exception - like the existing error "InvalidHeaderName" (An invalid character was found in header name)?

Comment (karelz) [reply src/libraries/Common/src/System/Net/Mail/MailBnfHelper.cs:272]: If we throw different exception, then we would have to just change the exception into another one in the caller IMO.

I didn't realize `ReadToken` is used directly, that makes things more complicated -- I am hesitant to start throwing different exception from the method without fully understanding all the usages and how it propagates from there. I think a proper fix should review all the usages of the parser methods and do the right thing IMO. Maybe adding comments and defining proper contract would be useful?

Comment (avanigupta) [reply src/libraries/Common/src/System/Net/Mail/MailBnfHelper.cs:272]: > If we throw different exception, then we would have to just change the exception into another one in the caller IMO.
> 

I found that in all callers ([except this one](https://github.com/dotnet/runtime/blob/fe41d3c762695cfeb364a8a0b61e2e376d22da5a/src/libraries/System.Net.Mail/src/System/Net/Mime/ContentType.cs#L113-L118)), we are already changing the exception in the caller.


I agree that we need a good contract definition for the parser. However, the scope of this PR is to fix the `IndexOutOfRangeException` thrown specifically because we are accessing `data[offset]` while throwing `FormatException` [here](https://github.com/dotnet/runtime/blob/fe41d3c762695cfeb364a8a0b61e2e376d22da5a/src/libraries/Common/src/System/Net/Mail/MailBnfHelper.cs#L272).

If changing the exception message doesn't look good, another option is to change the condition which throws this exception:
[`if (start == offset)`](https://github.com/dotnet/runtime/blob/fe41d3c762695cfeb364a8a0b61e2e376d22da5a/src/libraries/Common/src/System/Net/Mail/MailBnfHelper.cs#L270) can be changed to `if (start == offset && offset < data.Length)`

For more in depth changes to this code, I propose we create a new work item for a proper fix.

Comment (avanigupta) [reply src/libraries/Common/src/System/Net/Mail/MailBnfHelper.cs]: Hi @karelz , do these changes look good?

File: src/libraries/Common/src/System/Net/Mail/MailBnfHelper.cs
Comment (ManickaP) [inline src/libraries/Common/src/System/Net/Mail/MailBnfHelper.cs:270]: If `offset < data.Length` then the previous `for` loop must have ended in `else if (!Ttext[data[offset]])`.
Before this change, we would substring the data up until the first non-token character. And now it seems we're throwing?
This doesn't seem right.

**Review by ManickaP — APPROVED** [review]
LGTM

**Comment by avanigupta** [comment]
Hello, can this PR be closed?

**Comment by ManickaP** [comment]
> Hello, can this PR be closed?

@karelz you're blocking this PR. Do you have some more comments or are you happy with it and we can merge?

**Review by karelz — APPROVED** [review]
Sorry for the delay - I didn't realize I am blocking the PR. Feel free to merge.

**Pull request merged**
